idgenerator/idgenerator.go

```diff
@@ -0,0 +1,107 @@
+/*
+Package idgenerator contains several Span and Trace ID generators which can be
+used by the Zipkin tracer. Additional third party generators can be plugged in
+if they adhere to the IDGenerator interface.
+*/
+package idgenerator
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/openzipkin/zipkin-go/model"
+)
+
+var (
+	seededIDGen = rand.New(rand.NewSource(time.Now().UnixNano()))
+	// NewSource returns a new pseudo-random Source seeded with the given value.
+	// Unlike the default Source used by top-level functions, this source is not
+	// safe for concurrent use by multiple goroutines. Hence the need for a mutex.
+	seededIDLock sync.Mutex
+)
+
+// IDGenerator interface can be used to provide the Zipkin Tracer with custom
+// implementations to generate Span and Trace IDs.
+type IDGenerator interface {
+	SpanID() model.ID       // Generates a new Span ID
```

Comment:
eventhough you won't use it, almost certainly you'll want to pass parameters if generating a span ID.

Here's an old discussion (though we tossed the pluggability) https://github.com/openzipkin/brave/pull/110

main thing is that inside a trace, ids can be derived at least from the trace ID if not other things, eliminating some duplicate stuff. One way to avoid making a decision is to not make it pluggable until there's a use case.

Reply:
agreed... SpanID will be passed a non empty TraceID in case the generator is requested to generate a SpanID for a root span. This will allow the generator to just use the traceID as basis for the root spanID. This is less expensive then generating a separate one and will be appreciated in edge nodes like front-ends and gateways.